optuna/storages/_rdb/storage.py

```diff
@@ -446,46 +468,50 @@ def _create_new_trial(
 
         # Retry a couple of times. Deadlocks may occur in distributed environments.
         n_retries = 0
-        with _create_scoped_session(self.scoped_session) as session:
-            while True:
+        MAX_RETRIES = 5
+        while n_retries <= MAX_RETRIES:
+            # TODO(takizawa): Ponder if this exponential backoff is useful.
+            if n_retries != 0:
+                time.sleep((2**n_retries) * (1.0 + random.random()))
+            with _create_scoped_session(self.scoped_session, ignore_timeout_error=True) as session:
                 try:
                     # Ensure that that study exists.
                     #
                     # Locking within a study is necessary since the creation of a trial is not an
                     # atomic operation. More precisely, the trial number computed in
                     # `_get_prepared_new_trial` is prone to race conditions without this lock.
                     models.StudyModel.find_or_raise_by_id(study_id, session, for_update=True)
-
                     trial = self._get_prepared_new_trial(study_id, template_trial, session)
+                    if template_trial:
+                        frozen = copy.deepcopy(template_trial)
+                        frozen.number = trial.number
+                        frozen.datetime_start = trial.datetime_start
+                        frozen._trial_id = trial.trial_id
+                    else:
+                        frozen = FrozenTrial(
+                            number=trial.number,
+                            state=trial.state,
+                            value=None,
+                            values=None,
+                            datetime_start=trial.datetime_start,
+                            datetime_complete=None,
+                            params={},
+                            distributions={},
+                            user_attrs={},
+                            system_attrs={},
+                            intermediate_values={},
+                            trial_id=trial.trial_id,
+                        )
```

Comment:
I think these lines should be placed outside of the `while` loop. It would be better to use another scoped session, since these are read only steps.